### PR TITLE
[GOBBLIN-863]Handle race condition issue for hive registration

### PR DESCRIPTION
**Pull request by ZihanLi58**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-863

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
When updating the hive table, lock the table and fetch the latest schema from schema registry and use the latest schema to update table to make sure the table always has the latest schema

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Test streaming job to see hive registration work as expected. And fetching latest latest schema may cost 0.05~0.15 second. But this should not happen frequently, only happen when there if new partition or schema change.

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by ZihanLi58** [comment]
@sv2000 Can you take a look at this change? Thanks

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java [hunk 1]
```diff
@@ -93,9 +98,11 @@
   public static final String CREATE_HIVE_DATABASE = HIVE_REGISTER_METRICS_PREFIX + "createDatabaseTimer";
   public static final String CREATE_HIVE_TABLE = HIVE_REGISTER_METRICS_PREFIX + "createTableTimer";
   public static final String GET_HIVE_TABLE = HIVE_REGISTER_METRICS_PREFIX + "getTableTimer";
+  public static final String GET_AND_SET_LATEST_SCHEMA = HIVE_REGISTER_METRICS_PREFIX + "getAndSetLatestSchemaTimer";
   public static final String DROP_TABLE = HIVE_REGISTER_METRICS_PREFIX + "dropTableTimer";
   public static final String PATH_REGISTER_TIMER = HIVE_REGISTER_METRICS_PREFIX + "pathRegisterTimer";
   public static final String SKIP_PARTITION_DIFF_COMPUTATION = HIVE_REGISTER_METRICS_PREFIX + "skip.partition.diff.computation";
+  public static final String FETCH_LATEST_SCHEMA = HIVE_REGISTER_METRICS_PREFIX + "fetch.latest.schema";
```
Comment (sv2000) [inline gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java]: How about HIVE_REGISTER_METRICS_PREFIX + "fetchLatestSchemaFromSchemaRegistry"?

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java [hunk 2]
```diff
@@ -197,6 +210,16 @@ private boolean ensureHiveTableExistenceBeforeAlternation(String tableName, Stri
         try (Timer.Context context = this.metricContext.timer(GET_HIVE_TABLE).time()) {
           existingTable = HiveMetaStoreUtils.getHiveTable(client.getTable(dbName, tableName));
         }
+        if (this.schemaRegistry.isPresent()) {
+          try (Timer.Context context = this.metricContext.timer(GET_AND_SET_LATEST_SCHEMA).time()) {
+            String latestSchema = this.schemaRegistry.get().getLatestSchema(topicName).toString();
+            spec.getTable().getSerDeProps().setProp(AvroSerdeUtils.AvroTableProperties.SCHEMA_LITERAL.getPropName(), latestSchema);
+            table.getSd().setSerdeInfo(HiveMetaStoreUtils.getSerDeInfo(spec.getTable()));
+          } catch (SchemaRegistryException | IOException e) {
+            log.error(String.format("Error when fetch latest for topic %s", topicName), e);
```
Comment (sv2000) [inline gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java]: Minor typo: "Error when fetching latest schema for topic...".

File: gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java [hunk 3]
```diff
@@ -176,7 +189,7 @@ protected void registerPath(HiveSpec spec) throws IOException {
    * Or will create the table thru. RPC and return retVal from remote MetaStore.
    */
   private boolean ensureHiveTableExistenceBeforeAlternation(String tableName, String dbName, IMetaStoreClient client,
-      Table table, HiveSpec spec) throws TException{
+      Table table, HiveSpec spec) throws TException, IOException{
     try (AutoCloseableLock lock = this.locks.getTableLock(dbName, tableName)) {
```
Comment (sv2000) [inline gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/metastore/HiveMetaStoreBasedRegister.java]: Isn't this lock a JVM-local lock? Won't we need Hive Metastore side locking to serialize updates?

**Comment by autumnust** [comment]
LGTM, thanks for fixing.

**Pull request closed without merging**